testing/upload_validation.ts

```diff
@@ -480,12 +474,86 @@ const booleanPropertySchema = zodCompleteObject<BooleanSchema & ObjectSchemaProp
   ...basePropertyValidators,
 });
 
-const numberPropertySchema = zodCompleteObject<NumberSchema & ObjectSchemaProperty>({
+const baseNumberPropertySchema = zodCompleteStrictObject<BaseNumberSchema & ObjectSchemaProperty>({
+  type: zodDiscriminant(ValueType.Number),
+  ...basePropertyValidators,
+})
+
+const numericPropertySchema = baseNumberPropertySchema.extend({
+  codaType: zodDiscriminant(ValueHintType.Percent).optional(),
+  precision: z.number().optional(),
+  useThousandsSeparator: z.boolean().optional(),
+})
+
+const scalePropertySchema = baseNumberPropertySchema.extend({
+  type: zodDiscriminant(ValueType.Number),
+  codaType: zodDiscriminant(ValueHintType.Scale),
+  maximum: z.number(),
+  icon: z.nativeEnum(ScaleIconSet),
+  ...basePropertyValidators,
+})
+
+const sliderPropertySchema = baseNumberPropertySchema.extend({
+  type: zodDiscriminant(ValueType.Number),
+  codaType: zodDiscriminant(ValueHintType.Slider),
+  maximum: z.number().optional(),
+  minimum: z.number().optional(),
+  step: z.number().optional(),
+  ...basePropertyValidators,
+})
+
+const currencyPropertySchema = baseNumberPropertySchema.extend({
   type: zodDiscriminant(ValueType.Number),
-  codaType: z.enum([...NumberHintValueTypes]).optional(),
+  codaType: zodDiscriminant(ValueHintType.Currency),
+  precision: z.number().optional(),
+  currencyCode: z.string().optional(),
+  format: z.nativeEnum(CurrencyFormat).optional(),
   ...basePropertyValidators,
+})
+
+// TODO(alan): create base date, time, and datetime schemas to reuse across numeric schema
+// properties and string schema properties.
+
+const datePropertySchema = zodCompleteObject<DateSchema & ObjectSchemaProperty>({
```

Comment:
We need to define zod schemas for Date/Time/DateTime when the codaType is a string too, right, otherwise those fields will get stripped out when using a string formula?

Totally fine not to define a base zod schema, you can copy-paste the zod schemas but change the codaType to valueHintType.String.

Reply:
Yeah I guess I was trying to preserve the BaseDateSchema as much as possible.. will update!